src/analysis/processing/qgsalgorithmtransect.cpp

```diff
@@ -0,0 +1,186 @@
+         QObject::tr( "- TR_SEGMENT: ID of the segment of the linestring\n" ) +
+         QObject::tr( "- TR_ANGLE: Angle in degrees from the original line at the vertex\n" ) +
+         QObject::tr( "- TR_LENGTH: Total length of the transect returned\n" ) +
+         QObject::tr( "- TR_ORIENT: Side of the transect (only on the left or right of the line, or both side)\n" );
+
+}
+
+QgsTransectAlgorithm *QgsTransectAlgorithm::createInstance() const
+{
+  return new QgsTransectAlgorithm();
+}
+
+QVariantMap QgsTransectAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  Side orientation = static_cast< QgsTransectAlgorithm::Side >( parameterAsInt( parameters, QStringLiteral( "SIDE" ), context ) );
+  double angle = fabs( parameterAsDouble( parameters, QStringLiteral( "ANGLE" ), context ) );
+  double length = parameterAsDouble( parameters, QStringLiteral( "LENGTH" ), context );
+
+  if ( orientation == QgsTransectAlgorithm::Both )
+    length /= 2.0;
+
+  std::unique_ptr< QgsFeatureSource > source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !source )
+    return QVariantMap();
+
+  QgsFields fields = source->fields();
+
+  fields.append( QgsField( QStringLiteral( "TR_FID" ), QVariant::Int, QString(), 20 ) );
+  fields.append( QgsField( QStringLiteral( "TR_ID" ), QVariant::Int, QString(), 20 ) );
+  fields.append( QgsField( QStringLiteral( "TR_SEGMENT" ), QVariant::Int, QString(), 20 ) );
+  fields.append( QgsField( QStringLiteral( "TR_ANGLE" ), QVariant::Double, QString(), 5, 2 ) );
+  fields.append( QgsField( QStringLiteral( "TR_LENGTH" ), QVariant::Double, QString(), 20, 6 ) );
+  fields.append( QgsField( QStringLiteral( "TR_ORIENT" ), QVariant::Int, QString(), 1 ) );
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, fields,
+                                          QgsWkbTypes::LineString, source->sourceCrs() ) );
+  if ( !sink )
+    return QVariantMap();
+
+  QgsFeatureIterator features = source->getFeatures( );
+
+  int current = -1;
+  int number = 0;
+  double step =  source->featureCount() > 0 ? 100.0 / source->featureCount() : 1;
+  QgsFeature feat;
+
+
+  while ( features.nextFeature( feat ) )
+  {
+    current++;
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+
+    feedback->setProgress( current * step );
+    if ( !feat.hasGeometry() )
+      continue;
+
+    QgsGeometry inputGeometry = feat.geometry();
+
+    inputGeometry.convertToMultiType();
+    const QgsMultiLineString *multiLine = static_cast< QgsMultiLineString *  >( inputGeometry.geometry() );
+    for ( int id = 0; id < multiLine->numGeometries(); ++id )
+    {
+      const QgsLineString *line = static_cast< const QgsLineString * >( multiLine->geometryN( id ) );
+      QgsAbstractGeometry::vertex_iterator it = line->vertices_begin();
```

Comment:
@wonder-sk first use of your new iterator in the wild! win!

Reply:
Woohoo! And it looks like there is also a lot of scope for an iterator on geometry parts as well :-)